src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -301,23 +306,19 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
 
             return (IServiceProvider serviceProvider, object?[]? arguments) =>
             {
-                object?[]? constructorArguments = null;
-                if (parameters.Length != 0)
+                object?[] constructorArguments = new object?[parameters.Length];
+                for (int i = 0; i < parameters.Length; i++)
                 {
-                    constructorArguments = new object?[parameters.Length];
-                    for (int i = 0; i < parameters.Length; i++)
-                    {
-                        ref FactoryParameterContext parameter = ref parameters[i];
-                        constructorArguments[i] = ((parameter.ArgumentIndex != -1)
-                            // Throws an NullReferenceException if arguments is null. Consistent with expression-based factory.
-                            ? arguments![parameter.ArgumentIndex]
-                            : GetService(
-                                serviceProvider,
-                                parameter.ParameterType,
-                                declaringType,
-                                parameter.HasDefaultValue)) ?? parameter.DefaultValue;
+                    ref FactoryParameterContext parameter = ref parameters[i];
+                    constructorArguments[i] = ((parameter.ArgumentIndex != -1)
+                        // Throws an NullReferenceException if arguments is null. Consistent with expression-based factory.
```

Comment:
The expression-based implementation calls Expression.Convert on the argument: https://github.com/dotnet/runtime/blob/1bc528df807b3ed0c39143cbaab89b852a77b5a6/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L278

Are these conversions unnecessary here?

Reply:
I don't think so. When `ConstructorInfo.Invoke` is called with an array of arguments, doesn't it handle that for you?

Reply:
I tested:
```csharp
internal class ClassWithInt
{
    public int I { get; }
    public ClassWithInt(int i)
    {
        I = i;
    }
}

var factory1 = ActivatorUtilities.CreateFactory(typeof(ClassWithInt), new[] { typeof(int) });

var services = new ServiceCollection();
using var provider = services.BuildServiceProvider();
var item = (ClassWithInt)factory1(provider, new object[] { (byte)10 });
Assert.Equal(10, item.I);
```

It succeeds with new reflection factory.
It fails with old expressions factory:
> System.InvalidCastException: Unable to cast object of type 'System.Byte' to type 'System.Int32'.

Does `Invoke` have some extra logic around `IConvertable` values to make them work? How equal do we want to make these implementations?

Reply:
Yes, reflection `Invoke` does a number of bespoke conversions. The set of conversions done by reflection `Invoke` is different (neither subset or superset) from the set of conversions done by `Expression.Convert`.

I think it is very hard to make these implementations equal with the current reflection `Invoke`. I would ignore the problem for now and come back to it once we get to work on the new fast reflection Invoke APIs.